flytekit/image_spec/image_spec.py

```diff
@@ -0,0 +1,94 @@
+import base64
+import hashlib
+import json
+import os
+import sys
+from dataclasses import dataclass
+from typing import List, Optional
+
+import click
+from dataclasses_json import dataclass_json
+
+IMAGE_LOCK = f"{os.path.expanduser('~')}{os.path.sep}.flyte{os.path.sep}image.lock"
+
+
+@dataclass_json
+@dataclass
+class ImageSpec:
+    """
+    This class is used to specify the docker image that will be used to run the task.
+
+    Args:
+        registry: docker registry. if it's specified, flytekit will push the image.
+        packages: list of python packages that will be installed in the image.
+        apt_packages: list of ubuntu packages that will be installed in the image.
+        base_image: base image of the docker container.
+        python_version: python version in the image.
+        destination_dir: This is the location that the code should be copied into. This must be the same as the WORKING_DIR in the base image.
+    """
+
+    registry: Optional[str] = None
+    packages: Optional[List[str]] = None
+    apt_packages: Optional[List[str]] = None
+    base_image: Optional[str] = None
+    python_version: str = f"{sys.version_info.major}.{sys.version_info.minor}"
+    destination_dir: str = "/root"
+
+    def build_image(self, name: str, tag: str, fast_register: bool, source_root: Optional[str] = None):
+        raise NotImplementedError("This method is not implemented in the base class.")
+
+
+def build_docker_image(
+    image_spec: ImageSpec, name: str, tag: str, fast_register: bool, source_root: Optional[str] = None
+):
+    """
+    Build the docker image and push it to the registry.
+    """
+    if not image_exist(image_spec.registry, tag) and fast_register:
+        click.secho("The image has already been pushed. Skip building the image.", fg="blue")
+        return
+
+    image_spec.build_image(name, tag, fast_register, source_root)
+    update_lock_file(image_spec.registry, tag)
+
+
+def calculate_hash_from_image_spec(image_spec: ImageSpec):
+    """
+    Calculate the hash from the image spec.
+    """
+    h = hashlib.md5(bytes(image_spec.to_json(), "utf-8"))
+    tag = base64.urlsafe_b64encode(h.digest()).decode("ascii")
+    # replace "=" with "." to make it a valid tag
+    return tag.replace("=", ".")
+
+
+def image_exist(registry: str, tag: str) -> bool:
+    """
+    Check if the image has been pushed to the registry.
+    """
+    if os.path.isfile(IMAGE_LOCK) is False:
+        return True
```

Comment:
we decide to use docker sdk to check if the image exists in the registry. will remove lock file

Reply:
I think this is a good idea. I was very unsure whether I like the lockfile mechanism when reading it.

How will authentication work when using the python docker sdk? Is there a way to check the existence of a private image, without pulling it, in a way that the python client picks up the creds from `docker login`? Or does flytekit need to store credentials? I'm asking because in our internal registration helper script we have a function that retrieves the manifest of an image from the registry (with `requests.get`) in order to calculate the image digest. This helper requires an oauth token which is generated using a gcp auth library. This of course doesn't generalize well to other providers...

Reply:
> python client picks up the creds from docker login.

yes, so people don't need to worry about it if they use docker hub.

we'll support passing docker credentials to pyflyte command. 
```
pyflyte run --docker_cred token workflow.py wf
```

Reply:
> yes, so people don't need to worry about it if they use docker hub.

For instance on GCP, the `gcloud` sdk [can act as a credentials helper](https://cloud.google.com/sdk/gcloud/reference/auth/configure-docker) for docker for images stored in google container registry or the newer artefact registry. But if I understand correctly, for e.g. a gcr.io image, flytekit would defer to the docker python sdk, this one to docker, docker to the registered credentials helper, done :)